great_expectations/core/batch_manager.py

```diff
@@ -0,0 +1,169 @@
+import logging
+from collections import OrderedDict
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+from great_expectations.core.batch import (
+    Batch,
+    BatchData,
+    BatchDefinition,
+    BatchMarkers,
+)
+from great_expectations.core.id_dict import BatchSpec
+
+if TYPE_CHECKING:
+    from great_expectations.execution_engine import ExecutionEngine
+
+logger = logging.getLogger(__name__)
+logging.captureWarnings(True)
+
+
+class BatchManager:
+    def __init__(
+        self,
+        execution_engine: ExecutionEngine,
+        batch_list: Optional[List[Batch]] = None,
+    ) -> None:
+        """
+        Args:
+            execution_engine: The ExecutionEngine to be used to access cache of loaded Batch objects.
+            batch_list: List of Batch objects available from external source (default is None).
+        """
+        self._execution_engine = execution_engine
+
+        self._active_batch_id = None
+        self._active_batch_data_id = None
+
+        if batch_list is None:
+            batch_list = []
+
+        self._batch_cache = {}
+        self._batch_data_cache = {}
+
+        self.load_batch_list(batch_list=batch_list)
+        if len(batch_list) > 1:
+            logger.info(
+                f"{len(batch_list)} batches will be added to this Validator.  The batch_identifiers for the active "
+                f"batch are {self.active_batch.batch_definition['batch_identifiers'].items()}"
+            )
+
+    @property
+    def execution_engine(self) -> ExecutionEngine:
+        return self._execution_engine
+
+    @property
+    def batch_data_cache(self) -> Dict[str, BatchData]:
+        """Dictionary of loaded BatchData objects."""
+        return self._batch_data_cache
+
+    @property
+    def loaded_batch_ids(self) -> List[str]:
+        """IDs of loaded BatchData objects."""
+        return list(self._batch_data_cache.keys())
+
+    @property
+    def active_batch_data_id(self) -> Optional[str]:
+        """The batch id for the default batch data.
+
+        When a specific Batch objec is unavailable, then the data associated with the active_batch_data_id will be used.
+        """
+        if self._active_batch_data_id is not None:
+            return self._active_batch_data_id
+
+        if len(self._batch_data_cache) == 1:
+            return list(self._batch_data_cache.keys())[0]
+
+        return None
+
+    @property
+    def active_batch_data(self) -> Optional[BatchData]:
+        """The BatchData object from the currently-active Batch object."""
+        if self.active_batch_data_id is None:
+            return None
+
+        return self._batch_data_cache.get(self.active_batch_data_id)
+
+    @property
+    def batch_cache(self) -> Dict[str, Batch]:
+        """Getter for ordered dictionary (cache) of "Batch" objects in use (with batch_id as key)."""
+        if not isinstance(self._batch_cache, OrderedDict):
```

Comment:
In `__init__` we set `self._batch_cache` to be {}. Can we set to to an `OrderedDict` there to avoid this check here?

Also, we do want the return type of this property to be OrderedDict instead of dict?

As an aside, in python 3.7+, plain dictionaries are specified to preserve order. I'm not opposed to using an OrderedDict since it is a little esoteric but just wanted to mention it.

Reply:
Was about to ask the same question.
The minimum python for gx 3.7. So we don't really need an ordered dict... unless there some of the differences laid out here are important.
https://docs.python.org/3/library/collections.html#ordereddict-objects

>Some differences from [dict](https://docs.python.org/3/library/stdtypes.html#dict) still remain:

>1. The regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) was designed to be very good at mapping operations. Tracking insertion order was secondary.

>2. The [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) was designed to be good at reordering operations. Space efficiency, iteration speed, and the performance of update operations were secondary.

>3. The [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) algorithm can handle frequent reordering operations better than [dict](https://docs.python.org/3/library/stdtypes.html#dict). As shown in the recipes below, this makes it suitable for implementing various kinds of LRU caches.

>4. The equality operation for [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) checks for matching order.

>5. A regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) can emulate the order sensitive equality test with p == q and all(k1 == k2 for k1, k2 in zip(p, q)).

>6. The popitem() method of [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) has a different signature. It accepts an optional argument to specify which item is popped.

>7. A regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) can emulate OrderedDict’s od.popitem(last=True) with d.popitem() which is guaranteed to pop the rightmost (last) item.

>8. A regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) can emulate OrderedDict’s od.popitem(last=False) with (k := next(iter(d)), d.pop(k)) which will return and remove the leftmost (first) item if it exists.

>9. [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) has a move_to_end() method to efficiently reposition an element to an endpoint.

>10. A regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) can emulate OrderedDict’s od.move_to_end(k, last=True) with d[k] = d.pop(k) which will move the key and its associated value to the rightmost (last) position.

>11. A regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) does not have an efficient equivalent for OrderedDict’s od.move_to_end(k, last=False) which moves the key and its associated value to the leftmost (first) position.